### ISSUE TITLE
Getting the "initial cause" of an error

### ISSUE THREAD

**Issue by nicolasartman** [zaggino/z-schema#158]
I've noticed that, for slightly deep schemas, the "real" invalid field that caused the error is sometimes buried deep in `inner` error objects with incomplete paths, and I'm not sure how to distinguish it from the other `inner` errors.

For example, for a [schema snippet like this](https://gist.github.com/nicolasartman/57a2542b1bc6e9add8db), a failure to conform to the `pattern` on the innermost child item ([here](https://gist.github.com/nicolasartman/57a2542b1bc6e9add8db#file-schema-snippet-json-L89)) causes a bunch of `anyOf` errors and additional property errors.

I understand why the invalid pattern causes the cascading validation errors (the deepest doesn't match the schema, therefore its parent doesn't either, etc.), but how can I get the error that started it all, ideally with the full path to the field?

Thanks in advance!

**Comment by zaggino** [comment]
Can you post the full JSON of the validation result here?

**Comment by nicolasartman** [comment]
sure thing—here it is: https://gist.github.com/nicolasartman/01d4c1837d5692091047

**Comment by nicolasartman** [comment]
@zaggino sorry for the bother; any suggestions on how I might better identify the root cause of such validation errors?

**Comment by zaggino** [comment]
@nicolasartman reading the json is quite easy for me and there's no definite programming solution for what you want, ie getting the real error, lets dig into your report:

Base error does say that `navigation/items/1` `Data does not match any schemas from 'anyOf'` and from there you check inner which specifies a reason, why the data does not match, for every schema in `anyOf`

1. anyOf schema does not pass because `Additional properties not allowed: children`
2. anyOf schema does not pass because `Missing required property: path`
3. anyOf schema again includes `anyOf` so we again need to look at inner
3a. `String does not match pattern ^/([^/].*|$): //pattern-disallows-protocol-relative-urls-here`
3b. `Additional properties not allowed: path`
3c. `Missing required property: children`

Basically, fixing any one of 1,2,3a,3b,3c will make the validation pass. How should the validator know, which one is the "real" error?

**Comment by nicolasartman** [comment]
As far as I understand and have tried, fixing 1, 2, 3b, or 3c (anything except 3a) won't actually make the validation pass, because it'll still be failing other validations. For example, if one fixes 3b and removes the path property, 3c will still be present (and a new failure, a missing `path` property, will appear).

Reading the code, it appears to me that the the validator builds up nested lists recursively, then adds the missing anyOf error reports if a child fails to conform to any of the schemas. This suggests to me that the `pattern` match failure is likely the first actual error report object added to the nested list structure. That first item is what I'm after, since in all my tests that is most often the root cause (from the user's perspective) of the error.

I'm guessing that this is not possible or perhaps not a good design choice for the validator, so if that is the case then I'll try to handle that logic in my code by examining the error output.

**Comment by zaggino** [comment]
Well you could fix 1) by removing children and adding path property so it'd pass the first anyOf branch and the whole validation would succeed.

But you answered it. Doing any assumptions about the schema would be wrong for the validator so this is really not resolvable.
